lib/trino-hdfs/src/main/java/io/trino/hdfs/authentication/CachingKerberosHadoopAuthentication.java

```diff
@@ -30,21 +30,27 @@
 {
     private final KerberosHadoopAuthentication delegate;
 
-    @GuardedBy("this")
-    private UserGroupInformation userGroupInformation;
+    private final UserGroupInformation userGroupInformation;
     @GuardedBy("this")
     private long nextRefreshTime;
 
     public CachingKerberosHadoopAuthentication(KerberosHadoopAuthentication delegate)
     {
         this.delegate = requireNonNull(delegate, "delegate is null");
+        this.userGroupInformation = requireNonNull(delegate.getUserGroupInformation(), "delegate.getUserGroupInformation() is null");
+        nextRefreshTime = calculateNextRefreshTime(userGroupInformation);
     }
 
     @Override
     public synchronized UserGroupInformation getUserGroupInformation()
     {
-        if (nextRefreshTime < System.currentTimeMillis() || userGroupInformation == null) {
-            userGroupInformation = requireNonNull(delegate.getUserGroupInformation(), "delegate.getUserGroupInformation() is null");
+        if (nextRefreshTime < System.currentTimeMillis()) {
+            Subject existingSubject = getSubject(userGroupInformation);
+            UserGroupInformation newUserGroupInformation = requireNonNull(delegate.getUserGroupInformation(), "delegate.getUserGroupInformation() is null");
+            Subject newSubject = getSubject(newUserGroupInformation);
+            existingSubject.getPrincipals().addAll(newSubject.getPrincipals());
+            existingSubject.getPrivateCredentials().addAll(newSubject.getPrivateCredentials());
```

Comment:
Should we just add or replace?

Reply:
add, replace causes race (the issue I ran into with my attempt).

Reply:
old ones get cleaned up automatically by the JDK

Reply:
makes sense!